File: Projekt/server/akcie.py
# server/akcie.py
# Blueprint "Akcie" – reťazce a promo akcie + odporúčanie výroby
# - perzistentné uloženie do JSON (funguje hneď bez DB)
# - dashboard upozornenia 5 dní vopred
# - odporúčanie výroby 3 pracovné dni pred akciou (výpočet materiálov)
#
# Registrácia v app.py:
# from server.akcie import akcie_bp
# app.register_blueprint(akcie_bp, url_prefix="/api/kancelaria/akcie")

import os, json, datetime as dt
from flask import Blueprint, request, jsonify
import db_connector

akcie_bp = Blueprint("akcie", __name__)

# --- cesty k dátam (JSON) ---
BASE_DIR = os.path.dirname(os.path.abspath(__file__))          # .../Projekt/server
PROJECT_ROOT = os.path.normpath(os.path.join(BASE_DIR, ".."))  # .../Projekt
DATA_DIR = os.path.join(PROJECT_ROOT, "data")                  # .../Projekt/data

CHAINS_PATH = os.path.join(DATA_DIR, "retazce.json")
PROMOS_PATH = os.path.join(DATA_DIR, "akcie.json")


def _ensure_data_dir():
    if not os.path.isdir(DATA_DIR):
        os.makedirs(DATA_DIR, exist_ok=True)


def _load(path, default):
    _ensure_data_dir()
    if not os.path.exists(path):
        return default
    try:
        with open(path, "r", encoding="utf-8") as f:
            return json.load(f)
    except Exception:
        return default


def _save(path, data):
    _ensure_data_dir()
    with open(path, "w", encoding="utf-8") as f:
        json.dump(data, f, ensure_ascii=False, indent=2)


def _next_id(items):
    return (max([x["id"] for x in items], default=0) + 1) if items else 1


def _parse_date(s):
    # očakávame ISO "YYYY-MM-DD"
    return dt.date.fromisoformat(s)


def _date_str(d):
    return d.isoformat()


def _is_weekend(d):
    return d.weekday() >= 5  # 5=So, 6=Ne


def working_days_before(start_date: dt.date, days: int) -> dt.date:
    """Vracia dátum o N pracovných dní skôr (víkendy preskakuje)."""
    d = start_date
    left = days
    while left > 0:
        d = d - dt.timedelta(days=1)
        if not _is_weekend(d):
            left -= 1
    return d


def count_days_inclusive(a: dt.date, b: dt.date) -> int:
    return (b - a).days + 1


# ==========================
# Hooky na systém (recept, sklad, náklady) – uprav podľa DB
# ==========================

def get_product_meta(product_id: int):
    """
    Vytiahne názov a výrobné meta z DB.
    Priorita: products -> katalog_produktov -> produkty (fallback).
    Vracia: {"unit": "kg"|"ks", "piece_weight_g": int, "name": str, "ean": str, "is_produced": bool, "category": str}
    """
    try:
        # 1) products (hlavná tabuľka)
        row = db_connector.execute_query(
            """
            SELECT p.id, p.nazov, p.ean, p.je_vyroba, p.production_unit, p.piece_weight_g
            FROM products p
            WHERE p.id = %s
            LIMIT 1
            """,
            (product_id,), fetch='one'
        )

        # 2) centr. katalóg
        if not row:
            row = db_connector.execute_query(
                """
                SELECT id, nazov, ean, je_vyroba, NULL AS production_unit, NULL AS piece_weight_g
                FROM katalog_produktov
                WHERE id = %s
                LIMIT 1
                """,
                (product_id,), fetch='one'
            )

        # 3) úplný fallback (legacy tabuľka)
        if not row:
            row = db_connector.execute_query(
                """
                SELECT id, nazov, ean, je_vyroba, NULL AS production_unit, NULL AS piece_weight_g
                FROM produkty
                WHERE id = %s
                LIMIT 1
                """,
                (product_id,), fetch='one'
            )

        if row:
            pu = row.get('production_unit')
            unit = 'ks' if (pu == 1 or str(pu).lower() == 'ks') else 'kg'
            return {
                "unit": unit,
                "piece_weight_g": int(row.get('piece_weight_g') or 0),
                "name": row.get('nazov') or f"Produkt #{product_id}",
                "ean": row.get('ean') or "",
                "is_produced": bool(row.get('je_vyroba')),
                "category": ""
            }
    except Exception:
        pass  # nech to nespadne – fallback nižšie

    # Fallback mock – keby DB padla, UI beží
    return {
        "unit": "kg",
        "piece_weight_g": 0,
        "name": f"Produkt #{product_id}",
        "ean": "",
        "is_produced": True,
        "category": ""
    }


def get_recipe_items(product_id: int):
    """
    Vráti zoznam položiek receptu: [{"material_id": int, "nazov": str, "qty_per_100kg": float}]
    TODO: dopoj na tvoje receptové tabuľky (aktuálne prázdny zoznam).
    """
    return []


def get_material_stock(material_id: int):
    """
    Vráti stav na sklade v kg (float).
    TODO: napoj na sklad. Default 0.
    """
    return 0.0


def get_material_avg_cost(material_id: int):
    """
    Vráti priemernú nákupnú cenu materiálu €/kg.
    TODO: napoj na náklady. Default None (neznáme).
    """
    return None


def estimate_daily_sales(product_id: int, chain_id: int, unit: str):
    """
    Hrubý odhad denného predaja počas akcie.
    TODO: napoj na historické predaje. Default: 50 ks alebo 50 kg denne.
    """
    return 50.0


def chain_multiplier(chain_id: int):
    """
    Umožní zvýhodniť/zoslabiť predaj podľa reťazca (napr. COOP 1.2x).
    TODO: ak chceš, pridaj per-reťazec multiplikátor do retazce.json.
    """
    chains = _load(CHAINS_PATH, {"chains": []}).get("chains", [])
    ch = next((c for c in chains if c["id"] == chain_id), None)
    if not ch:
        return 1.0
    return float(ch.get("multiplier", 1.0))


def profit_estimate(product_id: int, qty_units: float, price_net: float, unit: str):
    """
    Odhad zisku: Tržba – Náklady.
    - pre "kg": cena je €/kg → revenue = qty_kg * price_net
    - pre "ks": cena je €/ks → revenue = qty_ks * price_net
    Náklady: ak je recept (is_produced=True), spočítame z materiálov (avg cost * množstvá).
             inak None (čerstvé mäso bez receptu -> skús získať priemernú cenu produktu, ak máš).
    """
    meta = get_product_meta(product_id)
    recipe = get_recipe_items(product_id) if meta.get("is_produced") else []
    revenue = qty_units * price_net

    if not recipe:  # čerstvé alebo recept nie je zadaný
        # TODO: ak máš priemernú nákupnú cenu produktu (€/kg alebo €/ks), použi ju:
        avg_cost_per_u = None  # napr. get_product_avg_cost(product_id)
        if avg_cost_per_u is None:
            return {"revenue": revenue, "cost": None, "profit": None, "note": "Chýba priemerná cena produktu."}
        cost = qty_units * avg_cost_per_u
        return {"revenue": revenue, "cost": cost, "profit": revenue - cost, "note": ""}

    # materiálové náklady z receptu
    if unit == "ks":
        kg_final = qty_units * (meta.get("piece_weight_g", 0) or 0) / 1000.0
    else:
        kg_final = qty_units

    total_cost = 0.0
    missing_costs = []
    for it in recipe:
        need_kg = (float(it.get("qty_per_100kg", 0)) / 100.0) * kg_final
        c = get_material_avg_cost(int(it["material_id"]))
        if c is None:
            missing_costs.append(it.get("nazov", f"MAT#{it['material_id']}"))
            continue
        total_cost += need_kg * float(c)

    if missing_costs:
        return {"revenue": revenue, "cost": None, "profit": None,
                "note": f"Chýba priemerná cena: {', '.join(missing_costs)}"}
    return {"revenue": revenue, "cost": total_cost, "profit": revenue - total_cost, "note": ""}


# ==========================
# API – Reťazce
# ==========================

@akcie_bp.route("/chains/list", methods=["POST"])
def chains_list():
    data = _load(CHAINS_PATH, {"chains": []})
    return jsonify({"chains": data.get("chains", [])})


@akcie_bp.route("/chains/add", methods=["POST"])
def chains_add():
    payload = request.get_json(silent=True) or {}
    name = (payload.get("name") or "").strip()
    multiplier = float(payload.get("multiplier") or 1.0)
    if not name:
        return jsonify({"ok": False, "error": "Zadaj názov reťazca."}), 400
    data = _load(CHAINS_PATH, {"chains": []})
    chains = data.get("chains", [])
    new = {
        "id": _next_id(chains),
        "name": name,
        "multiplier": multiplier,
        "created_at": dt.datetime.utcnow().isoformat()
    }
    chains.append(new)
    _save(CHAINS_PATH, {"chains": chains})
    return jsonify({"ok": True, "chain": new})


# ==========================
# API – Akcie
# ==========================

@akcie_bp.route("/list", methods=["POST"])
def promos_list():
    payload = request.get_json(silent=True) or {}
    upcoming_only = bool(payload.get("upcoming_only"))
    today = dt.date.today()
    promos = _load(PROMOS_PATH, {"promos": []}).get("promos", [])
    if upcoming_only:
        promos = [p for p in promos if _parse_date(p["date_to"]) >= today]
    return jsonify({"items": promos})


@akcie_bp.route("/add", methods=["POST"])
def promos_add():
    p = request.get_json(silent=True) or {}
    # očakávané polia: chain_id, product_id, price_net, date_from, date_to, note
    try:
        chain_id = int(p.get("chain_id"))
        product_id = int(p.get("product_id"))
        price_net = float(p.get("price_net"))
        dfrom = _parse_date(p.get("date_from"))
        dto   = _parse_date(p.get("date_to"))
    except Exception:
        return jsonify({"ok": False, "error": "Neplatné údaje akcie."}), 400

    note = (p.get("note") or "").strip()
    promos = _load(PROMOS_PATH, {"promos": []}).get("promos", [])
    new = {
        "id": _next_id(promos),
        "chain_id": chain_id,
        "product_id": product_id,
        "price_net": price_net,
        "date_from": _date_str(dfrom),
        "date_to": _date_str(dto),
        "note": note,
        "created_at": dt.datetime.utcnow().isoformat()
    }
    promos.append(new)
    _save(PROMOS_PATH, {"promos": promos})
    return jsonify({"ok": True, "promo": new})


# ==========================
# API – Dashboard upozornenia (5 dní vopred)
# ==========================

@akcie_bp.route("/dashboard", methods=["POST"])
def promos_dashboard():
    today = dt.date.today()
    horizon = today + dt.timedelta(days=5)
    promos = _load(PROMOS_PATH, {"promos": []}).get("promos", [])
    chains = {c["id"]: c for c in _load(CHAINS_PATH, {"chains": []}).get("chains", [])}
    out = []
    for p in promos:
        df = _parse_date(p["date_from"])
        dt_to = _parse_date(p["date_to"])
        # zobraz: ak začne do 5 dní, alebo už prebieha
        if (today <= df <= horizon) or (df <= today <= dt_to):
            meta = get_product_meta(int(p["product_id"]))
            chain = chains.get(p["chain_id"], {"name": "?"})
            msg = (
                f"Pozor: {df.strftime('%d.%m.%Y')}–{dt_to.strftime('%d.%m.%Y')} "
                f"prebieha akcia v {chain['name']} na {meta.get('name','produkt')} "
                f"za {p['price_net']:.2f} € bez DPH. Treba sa pripraviť."
            )
            out.append({
                "promotion_id": p["id"],
                "message": msg,
                "chain": chain["name"],
                "product": meta.get("name","produkt"),
                "from": p["date_from"],
                "to": p["date_to"],
                "price_net": p["price_net"]
            })
    return jsonify({"items": out})


# ==========================
# API – Odporúčanie výroby (3 pracovné dni pred)
# ==========================

@akcie_bp.route("/recommend", methods=["POST"])
def promo_recommend():
    payload = request.get_json(silent=True) or {}
    promo_id = int(payload.get("promotion_id", 0))
    promos = _load(PROMOS_PATH, {"promos": []}).get("promos", [])
    p = next((x for x in promos if x["id"] == promo_id), None)
    if not p:
        return jsonify({"ok": False, "error": "Akcia neexistuje."}), 404

    df = _parse_date(p["date_from"])
    dt_to = _parse_date(p["date_to"])
    chain_id = int(p["chain_id"])
    product_id = int(p["product_id"])
    price_net = float(p["price_net"])
    meta = get_product_meta(product_id)

    # odporúčaný začiatok výroby
    start_prod = working_days_before(df, 3)

    # odhad množstva (jednotka = kg alebo ks podľa produktu)
    base_daily = estimate_daily_sales(product_id, chain_id, meta.get("unit", "kg"))
    mult = chain_multiplier(chain_id)
    days = max(1, count_days_inclusive(df, dt_to))
    qty_units = base_daily * mult * days

    # materiály podľa receptu
    recipe = get_recipe_items(product_id)
    mats = []

    # finálne kg (ak ks → podľa hmotnosti kusu)
    if meta.get("unit") == "ks":
        final_kg = qty_units * (meta.get("piece_weight_g", 0)/1000.0)
    else:
        final_kg = qty_units

    for it in recipe:
        need_kg = (float(it.get("qty_per_100kg", 0))/100.0) * final_kg
        stock_kg = float(get_material_stock(int(it["material_id"])) or 0.0)
        buy_kg = max(0.0, need_kg - stock_kg)
        mats.append({
            "material_id": int(it["material_id"]),
            "nazov": it.get("nazov", f"MAT#{it['material_id']}"),
            "need_kg": round(need_kg, 3),
            "stock_kg": round(stock_kg, 3),
            "buy_kg": round(buy_kg, 3)
        })

    # zisk
    pe = profit_estimate(product_id, qty_units, price_net, meta.get("unit","kg"))

    return jsonify({
        "ok": True,
        "promotion": p,
        "product": {
            "id": product_id,
            "name": meta.get("name","produkt"),
            "unit": meta.get("unit","kg"),
            "piece_weight_g": meta.get("piece_weight_g", 0)
        },
        "recommendation": {
            "production_start": _date_str(start_prod),
            "qty_units": round(qty_units, 2),
            "qty_units_label": "kg" if meta.get("unit") == "kg" else "ks",
            "final_kg": round(final_kg, 3)
        },
        "materials": mats,
        "profit_estimate": pe
    })
# ==========================
# API – Vytvoriť plán výroby z akcie
# ==========================

@akcie_bp.route("/create_task", methods=["POST"])
def promo_create_task():
    payload = request.get_json(silent=True) or {}
    promo_id = int(payload.get("promotion_id") or 0)
    produce_date = (payload.get("produce_date") or "").strip()  # "YYYY-MM-DD"
    qty_units = float(payload.get("qty_units") or 0)

    if promo_id <= 0 or not produce_date or qty_units <= 0:
        return jsonify({"ok": False, "error": "Chýba promotion_id / dátum / množstvo > 0"}), 400

    # nájdi promo
    promos = _load(PROMOS_PATH, {"promos": []}).get("promos", [])
    promo = next((x for x in promos if x["id"] == promo_id), None)
    if not promo:
        return jsonify({"ok": False, "error": "Akcia neexistuje."}), 404

    product_id = int(promo["product_id"])
    meta = get_product_meta(product_id)

    # prepočty: uložíme plán v KG (výroba pracuje v kg)
    if meta.get("unit") == "ks":
        final_kg = qty_units * (meta.get("piece_weight_g", 0) or 0) / 1000.0
    else:
        final_kg = qty_units

    # datum_vyroby → spravíme datetime (napr. 06:00) kvôli typu stĺpca
    try:
        d = dt.datetime.strptime(produce_date, "%Y-%m-%d")
    except Exception:
        return jsonify({"ok": False, "error": "Neplatný formát dátumu (YYYY-MM-DD)"}), 400
    dt_prod = dt.datetime(d.year, d.month, d.day, 6, 0, 0)

    # INSERT do zaznamy_vyroba ako „planned“
    # (ak máš inú tabuľku pre úlohy výroby, zmeň SQL podľa seba)
    sql = """
        INSERT INTO zaznamy_vyroba
            (vyrobok_id, datum_vyroby, planovane_mnozstvo, skutocne_vyrobene, stav, celkova_cena_surovin)
        VALUES
            (%s, %s, %s, NULL, %s, NULL)
    """
    status = "planned"
    conn = None
    try:
        conn = db_connector.get_connection()
        cur = conn.cursor()
        cur.execute(sql, (product_id, dt_prod, final_kg, status))
        conn.commit()
        new_id = cur.lastrowid
    except Exception as e:
        if conn: conn.rollback()
        return jsonify({"ok": False, "error": f"DB chyba: {str(e)}"}), 500
    finally:
        try:
            if cur: cur.close()
        except: pass
        try:
            if conn and conn.is_connected(): conn.close()
        except: pass

    return jsonify({
        "ok": True,
        "task": {
            "id": new_id,
            "product_id": product_id,
            "product_name": meta.get("name",""),
            "plan_kg": round(final_kg, 3),
            "produce_date": produce_date,
            "status": status
        }
    })

File: Projekt/static/js/kancelaria_modules/akcie.js
// static/js/kancelaria_modules/akcie.js
// Modul "Akcie": reťazce + promo akcie + odporúčanie výroby + dashboard náhľad
// Registruje: window.initializeAkcieModule()

import { postJSON } from '/static/js/api_with_csrf.js?v=4';

const $   = (sel, r=document) => r.querySelector(sel);
const $$  = (sel, r=document) => Array.from(r.querySelectorAll(sel));
const on  = (el, ev, fn) => el && el.addEventListener(ev, fn);
const html = (el, h) => { if (el) el.innerHTML = h; };
const $id = (x)=> document.getElementById(x);

function ensureAkcieDom(){
  const host = $id('section-akcie');
  if (!host) return;
  if (host.innerHTML.trim()) return;

  html(host, `
    <h2>Akcie</h2>

    <div class="row gap">
      <button class="btn btn-tab active" id="akcie-tab-chains">Reťazce</button>
      <button class="btn btn-tab" id="akcie-tab-actions">Akcie</button>
      <button class="btn btn-tab" id="akcie-tab-dashboard">Nástenka</button>
    </div>

    <!-- Reťazce -->
    <div id="akcie-panel-chains" class="card mt">
      <h3>Reťazce</h3>
      <div class="row gap wrap">
        <input id="chain-name" placeholder="Názov reťazca (napr. COOP Jednota)">
        <input id="chain-mult" type="number" step="0.05" value="1.0" style="width:120px" title="koeficient predaja (1.0 = 100%)">
        <button id="chain-add" class="btn btn-primary">Pridať reťazec</button>
      </div>
      <p id="chain-msg" class="muted"></p>
      <div class="table-wrap mt">
        <table class="table">
          <thead><tr><th>ID</th><th>Názov</th><th>Koeficient</th><th>Vytvorené</th></tr></thead>
          <tbody id="chain-tbody"></tbody>
        </table>
      </div>
    </div>

    <!-- Akcie -->
    <div id="akcie-panel-actions" class="card mt" style="display:none">
      <h3>Vytvoriť akciu</h3>
      <div class="form-grid form-3">
        <label>Reťazec
          <select id="promo-chain"></select>
        </label>
        <label>Výrobok
          <select id="promo-product"></select>
        </label>
        <label>Predajná cena bez DPH (€)
          <input id="promo-price" type="number" step="0.01" min="0">
        </label>
        <label>Od dátumu
          <input id="promo-from" type="date">
        </label>
        <label>Do dátumu
          <input id="promo-to" type="date">
        </label>
        <label class="col-span-2">Poznámka
          <input id="promo-note" placeholder="voliteľné">
        </label>
      </div>
      <div class="row right mt">
        <button id="promo-add" class="btn btn-primary">Uložiť akciu</button>
      </div>
      <p id="promo-msg" class="muted"></p>

      <h3 class="mt">Zoznam akcií</h3>
      <div class="table-wrap">
        <table class="table">
          <thead><tr>
            <th>ID</th><th>Reťazec</th><th>Výrobok</th><th>Od</th><th>Do</th>
            <th>Cena (bez DPH)</th><th></th>
          </tr></thead>
          <tbody id="promo-tbody"></tbody>
        </table>
      </div>

      <div id="akcie-recommend" class="card mt" style="display:none"></div>
    </div>

    <!-- Nástenka (náhľad) -->
    <div id="akcie-panel-dashboard" class="card mt" style="display:none">
      <h3>Upozornenia (5 dní vopred)</h3>
      <ul id="dash-list" style="margin:.5rem 0 0 1rem"></ul>
    </div>
  `);
}

/* ====== pomocné API ====== */
async function apiChainsList(){ return postJSON('/api/kancelaria/akcie/chains/list', {}); }
async function apiChainsAdd(name, mult){ return postJSON('/api/kancelaria/akcie/chains/add', { name, multiplier: mult }); }
async function apiPromosList(upcoming_only=false){ return postJSON('/api/kancelaria/akcie/list', { upcoming_only }); }
async function apiPromosAdd(payload){ return postJSON('/api/kancelaria/akcie/add', payload); }
async function apiDashboard(){ return postJSON('/api/kancelaria/akcie/dashboard', {}); }
async function apiRecommend(promotion_id){ return postJSON('/api/kancelaria/akcie/recommend', { promotion_id }); }
async function apiCreateTask(promotion_id, produce_date, qty_units){
  return postJSON('/api/kancelaria/akcie/create_task', { promotion_id, produce_date, qty_units });
}

// z ERP – produkty (vyrobky). Môžeš vymeniť za iný endpoint s katalogom
async function apiErpProducts(){
  try{
    const r = await postJSON('/api/kancelaria/erp/recipes/products', {});
    return r?.products || [];
  }catch(_){ return []; }
}

/* ====== render ====== */
function renderChainsTable(items){
  const tb = $id('chain-tbody'); if (!tb) return;
  tb.innerHTML = '';
  items.forEach(c=>{
    const tr = document.createElement('tr');
    tr.innerHTML = `<td>${c.id}</td><td>${c.name}</td><td class="num">${(c.multiplier??1).toFixed(2)}</td><td>${c.created_at?.split('T')[0]||''}</td>`;
    tb.appendChild(tr);
  });
}

async function loadChainsIntoUI(){
  const data = await apiChainsList();
  const items = data?.chains || [];
  renderChainsTable(items);
  // select v akciách
  const sel = $id('promo-chain');
  if (sel) sel.innerHTML = items.map(c=>`<option value="${c.id}">${c.name}</option>`).join('');
}

async function loadProductsIntoUI(){
  const items = await apiErpProducts();
  const sel = $id('promo-product'); if (!sel) return;
  sel.innerHTML = items.map(p=>`<option value="${p.id}" data-unit="${p.production_unit===1?'ks':'kg'}" data-piece="${p.piece_weight_g||0}">${p.nazov}</option>`).join('');
}

function renderPromosTable(items, chainsLookup, productsLookup){
  const tb = $id('promo-tbody'); if (!tb) return;
  tb.innerHTML = '';
  items.forEach(p=>{
    const ch = chainsLookup.get(p.chain_id) || {name:'?'};
    const pr = productsLookup.get(p.product_id) || {nazov:`Produkt #${p.product_id}`};
    const tr = document.createElement('tr');
    tr.innerHTML = `
      <td>${p.id}</td>
      <td>${ch.name}</td>
      <td>${pr.nazov}</td>
      <td>${p.date_from}</td>
      <td>${p.date_to}</td>
      <td class="num">${Number(p.price_net).toFixed(2)}</td>
      <td class="right"><button class="btn" data-rec="${p.id}">Odporučiť výrobu</button></td>`;
    tr.querySelector('button[data-rec]').onclick = ()=> openRecommendation(p.id);
    tb.appendChild(tr);
  });
}

async function loadPromosIntoUI(){
  const [chains, products, promos] = await Promise.all([apiChainsList(), apiErpProducts(), apiPromosList(false)]);
  const chMap = new Map((chains?.chains||[]).map(c=>[c.id, c]));
  const prMap = new Map((products||[]).map(p=>[p.id, p]));
  renderPromosTable(promos?.items||[], chMap, prMap);
}

async function loadDashboardIntoUI(){
  const data = await apiDashboard();
  const ul = $id('dash-list'); if (!ul) return;
  ul.innerHTML = '';
  (data?.items||[]).forEach(x=>{
    const li = document.createElement('li');
    li.textContent = x.message;
    ul.appendChild(li);
  });
}

/* ====== recommendation panel ====== */
function renderRecommendation(res){
  const box = $id('akcie-recommend'); if (!box) return;
  if (!res?.ok){ box.style.display=''; box.innerHTML = `<p class="muted">Nepodarilo sa vypočítať odporúčanie.</p>`; return; }

  const meta = res.product || {};
  const rec  = res.recommendation || {};
  const pe   = res.profit_estimate || {};
  const mats = res.materials || [];

  box.style.display = '';
  box.innerHTML = `
    <h3>Odporúčanie výroby</h3>
    <div class="grid" style="display:grid;grid-template-columns:repeat(3,minmax(160px,1fr));gap:.5rem;">
      <div class="card">
        <div class="muted">Začať výrobu</div>
        <div><strong>${rec.production_start||'-'}</strong></div>
      </div>
      <div class="card">
        <div class="muted">Odporúčané množstvo</div>
        <div><strong>${rec.qty_units ?? '-'} ${rec.qty_units_label||''}</strong></div>
      </div>
      <div class="card">
        <div class="muted">Finálne množstvo (kg)</div>
        <div><strong>${rec.final_kg ?? '-'}</strong></div>
      </div>
    </div>

    <div class="card mt">
      <h4>Zadať do výroby</h4>
      <div class="row gap wrap">
        <label>Dátum výroby
          <input type="date" id="rec-prod-date" value="${rec.production_start||''}">
        </label>
        <label>Množstvo
          <input type="number" id="rec-qty" step="0.01" min="0" value="${(rec.qty_units ?? '')}">
        </label>
        <span class="muted">${meta.unit === 'ks' ? 'ks' : 'kg'}</span>
        <button id="rec-create-task" class="btn btn-primary">Zadať do výroby</button>
      </div>
      <p id="rec-msg" class="muted"></p>
    </div>

    <h4 class="mt">Materiály</h4>
    <div class="table-wrap">
      <table class="table">
        <thead><tr><th>Názov</th><th class="num">Potrebné (kg)</th><th class="num">Sklad (kg)</th><th class="num">Dokúpiť (kg)</th></tr></thead>
        <tbody>
          ${mats.map(m=>`<tr><td>${m.nazov}</td><td class="num">${Number(m.need_kg).toFixed(3)}</td><td class="num">${Number(m.stock_kg).toFixed(3)}</td><td class="num">${Number(m.buy_kg).toFixed(3)}</td></tr>`).join('')}
        </tbody>
      </table>
    </div>

    <h4 class="mt">Odhad zisku</h4>
    <p class="muted">${pe.note || ''}</p>
    <div class="grid" style="display:grid;grid-template-columns:repeat(3,minmax(160px,1fr));gap:.5rem;">
      <div class="card"><div class="muted">Tržba</div><div><strong>${pe.revenue != null ? pe.revenue.toFixed(2) + ' €' : '-'}</strong></div></div>
      <div class="card"><div class="muted">Náklady</div><div><strong>${pe.cost != null ? pe.cost.toFixed(2) + ' €' : '-'}</strong></div></div>
      <div class="card"><div class="muted">Zisk</div><div><strong>${pe.profit != null ? pe.profit.toFixed(2) + ' €' : '-'}</strong></div></div>
    </div>
  `;

  // zapoj tlačidlo „Zadať do výroby“
  $id('rec-create-task')?.addEventListener('click', async ()=>{
    const msg = $id('rec-msg'); if (msg) msg.textContent = '';
    const produce_date = $id('rec-prod-date')?.value || '';
    const qty = Number($id('rec-qty')?.value || 0);
    if (!produce_date || qty <= 0){
      msg && (msg.textContent = 'Zadaj dátum aj množstvo > 0.');
      return;
    }
    try{
      const r = await apiCreateTask(res.promotion.id || res.promotion_id || 0, produce_date, qty);
      msg && (msg.textContent = r?.ok ? '✅ Zadané do výroby.' : ('❌ ' + (r?.error || 'Chyba')));
    }catch(e){
      msg && (msg.textContent = '❌ ' + e.message);
    }
  });
}


async function openRecommendation(promotion_id){
  const res = await apiRecommend(promotion_id);
  renderRecommendation(res);
}

/* ====== wiring ====== */
let __wired = false;
function wireAkcieOnce(){
  if (__wired) return; __wired = true;

  // tabs
  document.addEventListener('click', (e)=>{
    const tabs = [
      {btn:'#akcie-tab-chains',    panel:'#akcie-panel-chains'},
      {btn:'#akcie-tab-actions',   panel:'#akcie-panel-actions'},
      {btn:'#akcie-tab-dashboard', panel:'#akcie-panel-dashboard'},
    ];
    for (const t of tabs){
      if (e.target?.matches(t.btn)){
        tabs.forEach(x=>{
          $id(x.btn.replace('#',''))?.classList.remove('active');
          const p = $id(x.panel.replace('#','')); if (p) p.style.display = 'none';
        });
        e.target.classList.add('active');
        const p = $id(t.panel.replace('#','')); if (p) p.style.display = '';

        if (t.panel==='#akcie-panel-chains') loadChainsIntoUI();
        if (t.panel==='#akcie-panel-actions') { loadChainsIntoUI(); loadProductsIntoUI(); loadPromosIntoUI(); }
        if (t.panel==='#akcie-panel-dashboard') loadDashboardIntoUI();
      }
    }
  });

  // reťazec
  on($id('chain-add'),'click', async ()=>{
    const name = ($id('chain-name')?.value || '').trim();
    const mult = Number($id('chain-mult')?.value || 1);
    const msg  = $id('chain-msg'); if (msg) msg.textContent = '';
    if (!name){ msg.textContent = 'Zadaj názov reťazca.'; return; }
    try{
      const r = await apiChainsAdd(name, mult);
      msg.textContent = r?.ok ? '✅ Uložené.' : ('❌ ' + (r?.error || 'Chyba'));
      await loadChainsIntoUI();
      $id('chain-name').value = '';
      $id('chain-mult').value = '1.0';
    }catch(e){ msg.textContent = '❌ ' + e.message; }
  });

  // akcia
  on($id('promo-add'),'click', async ()=>{
    const payload = {
      chain_id: Number($id('promo-chain')?.value || 0),
      product_id: Number($id('promo-product')?.value || 0),
      price_net: Number($id('promo-price')?.value || 0),
      date_from: $id('promo-from')?.value,
      date_to: $id('promo-to')?.value,
      note: $id('promo-note')?.value || ''
    };
    const msg = $id('promo-msg'); if (msg) msg.textContent='';
    if (!payload.chain_id || !payload.product_id || !payload.price_net || !payload.date_from || !payload.date_to){
      msg.textContent = 'Vyplň všetky povinné polia.'; return;
    }
    try{
      const r = await apiPromosAdd(payload);
      msg.textContent = r?.ok ? '✅ Akcia uložená.' : ('❌ ' + (r?.error || 'Chyba'));
      await loadPromosIntoUI();
    }catch(e){ msg.textContent = '❌ ' + e.message; }
  });
}

/* ====== public init ====== */
window.initializeAkcieModule = function(){
  if (window.__akcieInited) return;
  window.__akcieInited = true;

  const root = $id('section-akcie');
  if (!root){
    console.warn('[Akcie] Sekcia #section-akcie nie je v DOM.');
    return;
  }
  ensureAkcieDom();
  wireAkcieOnce();

  // default tabuľka
  $id('akcie-tab-chains')?.click();
};
